File: built-in plugins dsp/mb-harmonic-exciter.cpp
/**
 * MB Harmonic Exciter
 * Category : effect
 * Type     : distortion
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Multiband harmonic enhancer for adding presence and clarity
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_HARMONIC_EXCITER_H
#define MB_HARMONIC_EXCITER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbHarmonicExciter : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-harmonic-exciter";
    static constexpr const char* PLUGIN_NAME    = "MB Harmonic Exciter";
    static constexpr const char* PLUGIN_TYPE    = "distortion";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float low_amount = 0f;  // range [0, 100]
    float mid_amount = 0f;  // range [0, 100]
    float high_amount = 0f;  // range [0, 100]
    float crossover_low = 200f;  // range [50, 500]
    float crossover_high = 4000f;  // range [1000, 10000]
    float odd_even = 0.5f;  // range [0, 1]
    float mix = 0.5f;  // range [0, 1]
    };

    MbHarmonicExciter() = default;
    ~MbHarmonicExciter() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.low_amount = std::clamp(params.low_amount, 0f, 100f);
        params.mid_amount = std::clamp(params.mid_amount, 0f, 100f);
        params.high_amount = std::clamp(params.high_amount, 0f, 100f);
        params.crossover_low = std::clamp(params.crossover_low, 50f, 500f);
        params.crossover_high = std::clamp(params.crossover_high, 1000f, 10000f);
        params.odd_even = std::clamp(params.odd_even, 0f, 1f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Harmonic Exciter
        return input;
    }
};

#endif // MB_HARMONIC_EXCITER_H

File: built-in plugins dsp/mb-harmonic-exciter.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbHarmonicExciterPlugin: PluginDefinition = {
    id: 'mb-harmonic-exciter',
    slug: 'mb-harmonic-exciter',
    name: 'MB Harmonic Exciter',
    category: 'effect',
    type: 'distortion',
    version: '1.0.0',
    description: 'Multiband harmonic enhancer for adding presence and clarity',
    author: 'Max Booster',
    parameters: [
      { id: 'low_amount', name: 'Low Harmonics', type: 'float', defaultValue: 0, minValue: 0, maxValue: 100, unit: '%', automatable: true },
      { id: 'mid_amount', name: 'Mid Harmonics', type: 'float', defaultValue: 0, minValue: 0, maxValue: 100, unit: '%', automatable: true },
      { id: 'high_amount', name: 'High Harmonics', type: 'float', defaultValue: 0, minValue: 0, maxValue: 100, unit: '%', automatable: true },
      { id: 'crossover_low', name: 'Low Crossover', type: 'float', defaultValue: 200, minValue: 50, maxValue: 500, unit: 'Hz', automatable: true },
      { id: 'crossover_high', name: 'High Crossover', type: 'float', defaultValue: 4000, minValue: 1000, maxValue: 10000, unit: 'Hz', automatable: true },
      { id: 'odd_even', name: 'Odd/Even Balance', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true },
      { id: 'mix', name: 'Mix', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { low_amount: 0, mid_amount: 0, high_amount: 0, crossover_low: 200, crossover_high: 4000, odd_even: 0.5, mix: 0.5 },
  };

export default MbHarmonicExciterPlugin;
